### DISCUSSION TITLE
Procedure to contribute and add functionality

### DISCUSSION THREAD

**Discussion by chrhansen** [rheinwerk-verlag/pganonymize#21] [General]
Hey!

Thank you for this project.

I wanted to have some more functions. E.g.

- prepend (currently I only see append)
- More general string interpolation. E.g. provide and f-string in python or so.
- Be able to use values from other columns. E.g. if I use `fake.first_name` to replace a column that is otherwise a unique username I will have collisions using just `fake.first_name`. But if I could use `fake.first_name`-`primary-key` (I'd need more general string interpolation for this, I know), then I wouldn't have collisions. I _could_ use UUID to avoid collisions, but the UUID does not make for a convincing username.

Anyway, I didn't see any tests in the repo so I'm a bit worried making edits. What would you recommend?

Thanks!

**Comment by hkage** [comment]
Hey, 

thanks for your feedback and the suggestions. These are all features that make sense and would be nice additions to current feature set.

* A prepend feature would be easy to implement right now, although it is not implemented as a provider but within the `get_column_values` method. This is something I could  also implement in the next weeks or so.
* f-string notations would be really nice, although we have to keep the code Python 2 compatible right now. But I guess it would be possible to implement it manually, especially in addition to the column values (from the next point).
* I didn't think about re-using column values yet and how to make them accessible for the schema definition - as you already mentioned it, we would have to define a dedicated syntax to indicate we would like to reuse values from the current data row. That could make the usage of the provider classes more complex but I could imagine to provide a string interpolation for all providers (through the base class e.g.) so that it is possible to concatinate / interpolate values for all providers, no matter what their implementation looks like. But I would have to make some considerations first on how the syntax could look like.

The tests are indeed a weak spot of the project right now (I only started with some tests for the provider classes). I can only develop this project in my spare time and we currently are using simple bash SQL-statements for the anonymizing process. But we are planning to use the anonymizer, especially due to the increasing number of sensitive data to handle.

It would be great to have your suggestions being part of the project, but that would mean to add tests for at least those methods you would change for the features. Or you could wait for me to add tests for the existing methods, but honastly I cannot say right now how long that will take.

**Comment by hkage** [comment]
Hey,

just to mention it: with the pull request from https://github.com/rheinwerk-verlag/postgresql-anonymizer/pull/28 the code coverage for the tests dramatically increases and the contributor also added a string interpolation, like:

```yaml
format: '+65-{pga_value}-55-{id}'
```

`{pga_value}` contains the result from the used provider and `{id}` the values from the column.

So it might be worth taking a look into the next PyPi release that includes the PR.

Best regards!

**Reply by chrhansen** [reply]
Awesome. Thanks!
